File: Algorithm/Python/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python
#
# [34] Find First and Last Position of Element in Sorted Array
#

# @lc code=start
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        def find_idx(first_idx):
            left, right = 0, len(nums) - 1
            while left < right:
                mid = left + ((right - left)>>1)
                if nums[mid] == target:
                    if first_idx:
                        right = mid
                    else:
                        if mid + 1 < len(nums):
                            if nums[mid+1] != target:
                                left = mid
                                break
                            else:
                                left = mid + 1
                elif nums[mid] > target:
                    right = mid - 1
                else:
                    left = mid + 1
            
            return left if left < len(nums) and nums[left] == target else -1
        
        return [find_idx(True), find_idx(False)]
            

        
# @lc code=end

File: Algorithm/C++/34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start

#include <vector>

class Solution {
public:
    std::vector<int> searchRange(std::vector<int>& nums, int target) {
        return std::vector{findIdx(nums, target, true), findIdx(nums, target, false)};
    }

private:
    int findIdx(const std::vector<int>& nums, int target, bool first_idx) {
        int mid, left = 0, right = nums.size() - 1;
        if (left > right) {
            return -1;
        }
        while (left < right) {
            mid = left + ((right-left)>>1);
            if (nums[mid] == target) {
                if (first_idx) {
                    right = mid;
                } else {
                    if (mid + 1 < nums.size() && nums[mid+1] != target) {
                        left = mid;
                        break;
                    }
                    left = mid + 1;
                }
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        if (nums[left] != target) {
            return -1;
        } else {
            return left;
        }
    }
};
// @lc code=end

